src/game/Utils/Event_Pump.cc

```diff
@@ -10,62 +7,121 @@
 #include "Debug.h"
 #include "Logger.h"
 
-static void AddGameEventToQueue(UINT32 uiEvent, UINT16 usDelay, PTR pEventData, EventQueueID ubQueueID);
+#include <stdexcept>
+#include <vector>
 
+struct EVENT
+{
+	UINT32 TimeStamp;
+	UINT16 usDelay;
+	bool eventExpired;
+	GAMEEVENT gameEvent;
+};
 
-void AddGameEvent(GameEvent const uiEvent, UINT16 const usDelay, PTR const pEventData)
+enum class EventQueueID
 {
-	if (usDelay == DEMAND_EVENT_DELAY)
-	{
-		SLOGD("AddGameEvent: Sending Local #%d", uiEvent);
-		AddGameEventToQueue(uiEvent, 0, pEventData, DEMAND_EVENT_QUEUE);
-	}
-	else
-	{
-		SLOGD("AddGameEvent: Sending Local #%d", uiEvent);
-		AddGameEventToQueue(uiEvent, usDelay, pEventData, PRIMARY_EVENT_QUEUE);
-	}
+	PRIMARY_EVENT_QUEUE,
+	SECONDARY_EVENT_QUEUE,
+	DEMAND_EVENT_QUEUE
+};
+
+
+typedef std::vector<EVENT*>EventList;
+static EventList hEventQueue;
+static EventList hDelayEventQueue;
+static EventList hDemandEventQueue;
+
+
+static EventList& GetQueue(EventQueueID ubQueueID);
+
+
+static void AddEvent(GAMEEVENT const& gameEvent, UINT16 const usDelay, EventQueueID const ubQueueID)
+{
+	EVENT* pEvent = new EVENT{};
+	pEvent->TimeStamp  = GetJA2Clock();
+	pEvent->usDelay    = usDelay;
+	pEvent->eventExpired = false;
+	pEvent->gameEvent  = gameEvent;
+	GetQueue(ubQueueID).push_back(pEvent);
 }
 
 
-static void AddGameEventToQueue(UINT32 const uiEvent, UINT16 const usDelay, PTR const pEventData, EventQueueID const ubQueueID)
+static EVENT* RemoveEvent(UINT32 uiIndex, EventQueueID ubQueueID)
```

Comment:
Is this ever called with a `uiIndex != 0`? If not does it make sense to make this function a `PopEvent` function?

Reply:
I've never actually seen it happening but at least theoretically there is a code path in `DequeAllGameEvents()` where an expired event could be not the first in the secondary queue. But now that I've had a closer look at that code I think there is a bug here: if that ever happens we'd have an out of bound access to the vector because uiQueueSize is not updated. I'll post a fix tomorrow.

Reply:
At least theoretically there is a code path in `DequeAllGameEvents` where this could happen if an expired event is not the first in the secondary queue. But now that I've looked closer at that code I've noticed that we'd have an out of bounds access to the vector in that case. I'll post a fix later.